File: esp32_controller/stellarium.py
# stellarium.py - Stellarium Telescope Protocol server

import socket
import struct
import time
import _thread

from config import STELLARIUM_PORT, OBSERVER_LAT, OBSERVER_LON
from coordinates import alt_az_to_ra_dec
import main as app  # Access global state


def start_stellarium_server():
    """Start TCP server for Stellarium telescope protocol"""
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server.bind(('0.0.0.0', STELLARIUM_PORT))
    server.listen(1)

    print(f"Stellarium server listening on port {STELLARIUM_PORT}")

    while True:
        try:
            client, addr = server.accept()
            print(f"Stellarium connected from {addr}")
            _thread.start_new_thread(handle_stellarium_client, (client,))
        except Exception as e:
            print(f"Stellarium accept error: {e}")
            time.sleep(1)


def handle_stellarium_client(client):
    """Handle a Stellarium client connection"""
    try:
        while True:
            # Try to read a command
            data = client.recv(20)
            if not data:
                break

            if len(data) >= 20:
                # Parse goto command
                msg_len, msg_type = struct.unpack('<HH', data[0:4])

                if msg_type == 0:  # Goto command
                    # Stellarium sends RA/Dec as unsigned 32-bit values
                    # RA: 0x00000000 to 0x100000000 = 0h to 24h
                    # Dec: 0x00000000 to 0x100000000 = -90° to +90° (with 0x80000000 = 0°)
                    ra_raw, dec_raw = struct.unpack('<II', data[12:20])

                    # Convert to hours and degrees
                    ra_hours = ra_raw * 24.0 / 0x100000000
                    dec_deg = (dec_raw * 180.0 / 0x100000000) - 90.0

                    print(f"Stellarium goto: RA={ra_hours:.4f}h, Dec={dec_deg:.4f}°")

                    # Update global target
                    app.current_ra = ra_hours
                    app.current_dec = dec_deg
                    app.tracking_enabled = True

            # Send current position back to Stellarium
            send_position_to_stellarium(client)

            time.sleep(0.1)

    except Exception as e:
        print(f"Stellarium client error: {e}")
    finally:
        client.close()
        print("Stellarium client disconnected")


def send_position_to_stellarium(client):
    """Send current telescope position to Stellarium"""
    # Convert current Alt/Az back to RA/Dec for Stellarium
    ra_hours, dec_deg = alt_az_to_ra_dec(
        app.target_alt, app.target_az,
        OBSERVER_LAT, OBSERVER_LON
    )

    # Convert to Stellarium format
    ra_raw = int((ra_hours / 24.0) * 0x100000000) & 0xFFFFFFFF
    dec_raw = int(((dec_deg + 90.0) / 180.0) * 0x100000000) & 0xFFFFFFFF

    # Build response message
    # Length: 24 bytes, Type: 0, Time: current microseconds, RA, Dec, Status
    current_time = time.ticks_us()
    msg = struct.pack('<HHQIIi',
        24,           # message length
        0,            # message type
        current_time, # timestamp (microseconds)
        ra_raw,       # RA
        dec_raw,      # Dec
        0             # status
    )

    try:
        client.send(msg)
    except Exception as e:
        print(f"Error sending to Stellarium: {e}")

File: esp32_controller_arduino/src/stellarium.cpp
// stellarium.cpp - Stellarium Telescope Protocol server (Async version)

#include "stellarium.h"
#include "config.h"
#include "settings.h"
#include "state.h"
#include "coordinates.h"
#include "srt_serial.h"
#include <AsyncTCP.h>

static AsyncServer* stellariumServer = nullptr;
static AsyncClient* stellariumClient = nullptr;
static unsigned long lastPositionSend = 0;

extern SRTState state;

static void prepareStellariumTrackingTarget() {
    state.waitingForWrap = false;
    state.waitingForRise = false;
    state.movementHoldUntil = 0;
    if (state.azOnlyTracking) {
        state.azOnlyAlt = srtSerial.getCurrentAlt();
    }
    if (state.altOnlyTracking) {
        state.altOnlyAz = srtSerial.getCurrentAz();
    }
    state.trackingEnabled = true;
}

void sendPositionToStellarium() {
    if (!stellariumClient || !stellariumClient->connected()) return;

    // Convert current Alt/Az back to RA/Dec for Stellarium
    double raHours, decDeg;
    altAzToRaDec(state.targetAlt, state.targetAz, settings.observerLat, settings.observerLon, raHours, decDeg);

    // Convert to Stellarium format
    uint32_t raRaw = (uint32_t)((raHours / 24.0) * 0x100000000ULL) & 0xFFFFFFFF;
    int32_t decSigned = (int32_t)((decDeg / 90.0) * 0x40000000);

    // Build response message (24 bytes)
    uint8_t msg[24];
    uint16_t msgLen = 24;
    uint16_t msgType = 0;
    uint64_t timestamp = (uint64_t)micros();

    // Pack in little-endian format
    msg[0] = msgLen & 0xFF;
    msg[1] = (msgLen >> 8) & 0xFF;
    msg[2] = msgType & 0xFF;
    msg[3] = (msgType >> 8) & 0xFF;

    for (int i = 0; i < 8; i++) {
        msg[4 + i] = (timestamp >> (i * 8)) & 0xFF;
    }

    msg[12] = raRaw & 0xFF;
    msg[13] = (raRaw >> 8) & 0xFF;
    msg[14] = (raRaw >> 16) & 0xFF;
    msg[15] = (raRaw >> 24) & 0xFF;

    msg[16] = decSigned & 0xFF;
    msg[17] = (decSigned >> 8) & 0xFF;
    msg[18] = (decSigned >> 16) & 0xFF;
    msg[19] = (decSigned >> 24) & 0xFF;

    msg[20] = 0;
    msg[21] = 0;
    msg[22] = 0;
    msg[23] = 0;

    stellariumClient->write((const char*)msg, 24);
}

void onStellariumData(void* arg, AsyncClient* client, void* data, size_t len) {
    if (len >= 20) {
        uint8_t* bytes = (uint8_t*)data;

        uint16_t msgLen = bytes[0] | (bytes[1] << 8);
        uint16_t msgType = bytes[2] | (bytes[3] << 8);

        if (msgType == 0) {  // Goto command
            uint32_t raRaw = bytes[12] | (bytes[13] << 8) | (bytes[14] << 16) | (bytes[15] << 24);
            int32_t decSigned = bytes[16] | (bytes[17] << 8) | (bytes[18] << 16) | (bytes[19] << 24);

            double raHours = (double)raRaw * 24.0 / 4294967296.0;
            double decDeg = (double)decSigned * 90.0 / 1073741824.0;

            Serial.printf("Stellarium goto: RA=%.4fh, Dec=%.4f\n", raHours, decDeg);
            char logBuf[48];
            snprintf(logBuf, sizeof(logBuf), "Stellarium: RA=%.3fh Dec=%.1f", raHours, decDeg);
            srtSerial.logESP(logBuf);

            state.currentRA = raHours;
            state.currentDec = decDeg;
            state.targetName = "";
            prepareStellariumTrackingTarget();
        }
    }
}

void onStellariumClient(void* arg, AsyncClient* client) {
    Serial.println("Stellarium client connected");
    srtSerial.logESP("Stellarium connected");

    if (stellariumClient && stellariumClient != client) {
        // Disconnect old client
        stellariumClient->close();
    }

    stellariumClient = client;

    client->onData(onStellariumData, nullptr);

    client->onDisconnect([](void* arg, AsyncClient* c) {
        Serial.println("Stellarium client disconnected");
        srtSerial.logESP("Stellarium disconnected");
        if (stellariumClient == c) {
            stellariumClient = nullptr;
        }
        delete c;
    }, nullptr);

    client->onError([](void* arg, AsyncClient* c, int8_t error) {
        Serial.printf("Stellarium error: %d\n", error);
    }, nullptr);
}

void setupStellariumServer() {
    stellariumServer = new AsyncServer(STELLARIUM_PORT);
    stellariumServer->onClient(onStellariumClient, nullptr);
    stellariumServer->begin();
    Serial.printf("Stellarium async server listening on port %d\n", STELLARIUM_PORT);
}

void handleStellariumServer() {
    // Send position updates every 100ms if client connected
    if (stellariumClient && stellariumClient->connected()) {
        if (millis() - lastPositionSend >= 100) {
            sendPositionToStellarium();
            lastPositionSend = millis();
        }
    }
}
